File: test/markdown-render.test.ts
import { describe, expect, it } from 'vitest';
import { buildCleanCard, extractCardFences, renderRichText } from '../src/card/markdown-render';

/** Collect every element with a given tag from a card body / element list. */
function tags(node: unknown, tag: string, acc: any[] = []): any[] {
  if (Array.isArray(node)) node.forEach((n) => tags(n, tag, acc));
  else if (node && typeof node === 'object') {
    const o = node as Record<string, unknown>;
    if (o.tag === tag) acc.push(o);
    for (const k of Object.keys(o)) tags(o[k], tag, acc);
  }
  return acc;
}

describe('extractCardFences', () => {
  it('pulls the ```feishu-card fence and strips it from the text', () => {
    const text = '前言\n\n```feishu-card\n# 标题\n正文\n```\n\n后记';
    const { fences, stripped } = extractCardFences(text);
    expect(fences).toEqual(['# 标题\n正文']);
    expect(stripped).not.toContain('feishu-card');
    expect(stripped).toContain('前言');
    expect(stripped).toContain('后记');
  });

  it('handles multiple fences and leaves plain text untouched', () => {
    const plain = '没有卡片，只有 `code` 和 **bold**。';
    expect(extractCardFences(plain)).toEqual({ fences: [], stripped: plain });

    const two = '```feishu-card\nA\n```\nmid\n```feishu-card\nB\n```';
    const { fences } = extractCardFences(two);
    expect(fences).toEqual(['A', 'B']);
  });
});

describe('renderRichText', () => {
  it('plain text → a single markdown element (fast path)', () => {
    const els = renderRichText('就是一段**普通**文字');
    expect(els).toHaveLength(1);
    expect(els[0]).toMatchObject({ tag: 'markdown', content: '就是一段**普通**文字' });
  });

  it('interleaves text and uploaded images, splitting at the ref', () => {
    const map = new Map([['shot.png', 'img_key_123']]);
    const els = renderRichText('看这张图：\n\n![管理台](shot.png)\n\n然后呢', map);
    expect(els.map((e: any) => e.tag)).toEqual(['markdown', 'img', 'markdown']);
    const img = tags(els, 'img')[0];
    expect(img.img_key).toBe('img_key_123');
    expect(img.alt.content).toBe('管理台');
    expect((els[0] as any).content).toContain('看这张图');
    expect((els[2] as any).content).toContain('然后呢');
  });

  it('keeps the literal markdown when the image is unresolved (no upload)', () => {
    const els = renderRichText('图：![x](missing.png) 完', new Map());
    expect(tags(els, 'img')).toHaveLength(0);
    expect((els[0] as any).content).toContain('![x](missing.png)');
  });

  it('resolves http(s) URLs the same way (keyed by the verbatim src)', () => {
    const url = 'https://example.com/a.png';
    const els = renderRichText(`![](${url})`, new Map([[url, 'img_remote']]));
    expect(tags(els, 'img')[0].img_key).toBe('img_remote');
  });

  it('drops a ```feishu-card fence from the answer (hoisted to a clean card)', () => {
    const els = renderRichText('答复：\n\n```feishu-card\n# T\nbody\n```');
    // only the lead-in text remains; the fence is gone
    expect(tags(els, 'markdown').every((m: any) => !m.content.includes('feishu-card'))).toBe(true);
    expect((els[0] as any).content).toContain('答复');
  });

  it('preserves a Feishu mention tag verbatim (so codex can @ a user)', () => {
    const els = renderRichText('已处理完，请验收 <at id=ou_abcd1234></at>');
    expect(els).toHaveLength(1);
    expect((els[0] as any).content).toBe('已处理完，请验收 <at id=ou_abcd1234></at>');
  });

  it('preserves a mention even when interleaved with an uploaded image', () => {
    const map = new Map([['shot.png', 'img_key_1']]);
    const els = renderRichText('看图 <at id=ou_x></at>\n\n![s](shot.png)', map);
    expect((els[0] as any).content).toContain('<at id=ou_x></at>');
    expect(tags(els, 'img')).toHaveLength(1);
  });

  it('preserves a bare mention with no surrounding text', () => {
    const els = renderRichText('<at id=ou_x></at>');
    expect(els).toHaveLength(1);
    expect((els[0] as any).content).toBe('<at id=ou_x></at>');
  });
});

describe('buildCleanCard', () => {
  it('hoists the leading heading into a blue header and maps blocks', () => {
    const md = [
      '# 更新说明',
      '',
      '本次更新重点提升了体验。',
      '',
      '![管理台](admin.png)',
      '',
      '---',
      '',
      '**新增功能**',
      '- A',
      '- B',
      '',
      '> 一句话总结：更顺手了。',
    ].join('\n');
    const card: any = buildCleanCard(md, new Map([['admin.png', 'img_admin']]));

    expect(card.schema).toBe('2.0');
    expect(card.header.title.content).toBe('更新说明');
    expect(card.header.template).toBe('blue');
    expect(card.config.summary.content).toBe('更新说明');

    const els = card.body.elements as any[];
    expect(tags(els, 'img')[0].img_key).toBe('img_admin');
    expect(tags(els, 'hr')).toHaveLength(1);
    // the quote block becomes a grey note (div with grey lark_md), not markdown
    const note = els.find((e) => e.tag === 'div' && e.text?.text_color === 'grey');
    expect(note.text.content).toContain('一句话总结');
  });

  it('a title-only fence still yields a valid non-empty body', () => {
    const card: any = buildCleanCard('# 只有标题');
    expect(card.header.title.content).toBe('只有标题');
    expect((card.body.elements as any[]).length).toBeGreaterThan(0);
  });

  it('no leading heading → no header bar', () => {
    const card: any = buildCleanCard('直接正文，没有标题。');
    expect(card.header).toBeUndefined();
    expect((card.body.elements as any[])[0].content).toContain('直接正文');
  });
});

File: internal/card/markdown_render_test.go
package card

import (
	"strings"
	"testing"
)

func TestExtractCardFences(t *testing.T) {
	text := "before\n```feishu-card\n# Title\nbody\n```\nafter"
	r := ExtractCardFences(text)
	if len(r.Fences) != 1 || strings.TrimSpace(r.Fences[0]) != "# Title\nbody" {
		t.Fatalf("fences wrong: %+v", r.Fences)
	}
	if strings.Contains(r.Stripped, "feishu-card") || strings.Contains(r.Stripped, "Title") {
		t.Fatalf("stripped should remove fence: %q", r.Stripped)
	}
	if !strings.Contains(r.Stripped, "before") || !strings.Contains(r.Stripped, "after") {
		t.Fatalf("stripped should keep surrounding text: %q", r.Stripped)
	}
}

func TestRenderRichText_PlainShortCircuit(t *testing.T) {
	els := RenderRichText("just plain text", nil)
	if len(els) != 1 || els[0]["tag"] != "markdown" || els[0]["content"] != "just plain text" {
		t.Fatalf("plain short-circuit wrong: %+v", els)
	}
	if RenderRichText("   ", nil) != nil {
		t.Fatal("whitespace-only should return nil")
	}
}

func TestRenderRichText_ImageReplaced(t *testing.T) {
	text := "see ![a](p1.png) here"
	els := RenderRichText(text, map[string]string{"p1.png": "img_key_1"})
	// 应含一个 img + 前后文本 md。
	hasImg := false
	for _, e := range els {
		if e["tag"] == "img" && e["img_key"] == "img_key_1" {
			hasImg = true
		}
	}
	if !hasImg {
		t.Fatalf("image should be replaced with img element: %+v", els)
	}
}

func TestRenderRichText_UnresolvedImageKeptLiteral(t *testing.T) {
	text := "see ![a](nonexistent.png) here"
	els := RenderRichText(text, nil)
	// 未解析 → 保留原文（不静默丢）。
	joined := ""
	for _, e := range els {
		if c, ok := e["content"].(string); ok {
			joined += c
		}
	}
	if !strings.Contains(joined, "nonexistent.png") {
		t.Fatalf("unresolved image should keep literal markdown: %q", joined)
	}
}

func TestBuildCleanCard_TitleHeader(t *testing.T) {
	c := BuildCleanCard("# 我的卡片\nbody line\n\n> note", nil, HeaderBlue)
	h, ok := c["header"]
	if !ok {
		t.Fatal("title should become header")
	}
	if h.(CardElement)["title"].(CardElement)["content"] != "我的卡片" {
		t.Fatal("header title wrong")
	}
	body := c["body"].(CardElement)
	els := body["elements"].([]CardElement)
	if len(els) == 0 {
		t.Fatal("body should have elements")
	}
	// 应含 note（> 引用）。
	hasNote := false
	for _, e := range els {
		if e["tag"] == "div" {
			hasNote = true
		}
	}
	if !hasNote {
		t.Fatalf("quote block should map to note: %+v", els)
	}
}

func TestBuildCleanCard_NoTitle(t *testing.T) {
	c := BuildCleanCard("just body", nil, HeaderBlue)
	if _, ok := c["header"]; ok {
		t.Fatal("no heading → no header")
	}
}

func TestBuildCleanCard_HrBlock(t *testing.T) {
	c := BuildCleanCard("# T\n\n---\n\nbody", nil, HeaderBlue)
	body := c["body"].(CardElement)
	els := body["elements"].([]CardElement)
	hasHr := false
	for _, e := range els {
		if e["tag"] == "hr" {
			hasHr = true
		}
	}
	if !hasHr {
		t.Fatalf("--- should map to hr: %+v", els)
	}
}
